src/imitation/data/serialize.py

```diff
@@ -0,0 +1,178 @@
+"""Serialization utilities for trajectories."""
+import logging
+import os
+import pathlib
+import warnings
+from typing import Mapping, Optional, Sequence, cast
+
+import datasets
+import numpy as np
+
+from imitation.data import huggingface_datasets_conversion as hfds
```

Comment:
Hmmm... Maybe have a shorter module name in the first place?

Reply:
We could go with `hf_datasets_conversion`? I am happy about other suggestions but I would like the name to stay descriptive. I rather type some more letters now than wonder what a module was supposed to do later.

Reply:
I might call it huggingface_data_converter or something like that? I think having a very descriptive name but then renaming the import to a non-obvious abbreviation is probably just as bad for redability

Reply:
Hmm. I am not really happy with that name either. I would expect a class named `HuggingFaceDataConverter` in such a module.

When it is so hard to come up with a good name for a module that is often an indicator for bad architecture of the module. So I looked at its content again and the following came up:

When saving, we create a HF dataset just to write it to disk immediately. This involves a (shallow) copy of the data.

When loading, we have a wrapper that makes a HF dataset visible as a sequence of trajectories. Trajectory objects generated on-the-fly.

To make this symmetrical, I should have used the `Dataset.from_generator` method instead of the `Dataset.from_dict` method with a generator that constructs dicts from trajectories on-the-fly ([documentation here](https://huggingface.co/docs/datasets/v2.10.0/en/create_dataset#from-local-files)).
This is also the most memory-efficient way to create datasets and it would make it possible to stream sampled trajectories straight to disk.
I will push a version of this soon.

Concerning the naming of the module: I would propose to just call it `huggingface_utils` or even `hf_utils`. In symmetry to that we could refactor `imitation.policies.serialize` and pull out most of the HF-specific code in a `imitation.policies.huggingface_utils` module.

Reply:
That sounds like a good decision, thanks for taking a look at this! Huggingface_utils sounds good.

Reply:
Added it just now. Unfortunately this change made the save a lot slower (`test_types.test_save_trajectories` executes orders of magnitude slower). Maybe here @simoninithomas can give us some insight?

Reply:
So I asked the dataset team, that can explain the problem:
- `from_dict` writes in RAM
- `from_generator` writes on disk

Reply:
So should I better switch between `from_dict` and `from_generator` based on the size of the dataset (assuming that I know the number of elements coming out of the generator) or is that something, that you maybe implement on the datasets library end?

Reply:
If your dataset is a python dict, it takes up RAM and is therefore relatively small so you can load it with from_dict. But if you both load files one by one, then from_generator is better